File: web-hrd/lib/services/leaveService.js
const API_URL = process.env.NEXT_PUBLIC_API_URL;

const getHeaders = () => {
  const token = localStorage.getItem("token");
  return {
    "Content-Type": "application/json",
    Authorization: `Bearer ${token}`,
  };
};

export const leaveService = {
  getLeaves: async ({ month, year } = {}) => {
    const params = new URLSearchParams();
    if (month) params.set("month", month);
    if (year) params.set("year", year);
    const query = params.toString() ? `?${params.toString()}` : "";
    const response = await fetch(`${API_URL}/leaves${query}`, {
      headers: getHeaders(),
    });
    return response.json();
  },
  applyLeave: async (data) => {
    const response = await fetch(`${API_URL}/leaves`, {
      method: "POST",
      headers: getHeaders(),
      body: JSON.stringify(data),
    });
    return response.json();
  },
  updateStatus: async (id, status) => {
    const response = await fetch(`${API_URL}/leaves/${id}/status`, {
      method: "PUT",
      headers: getHeaders(),
      body: JSON.stringify({ status }),
    });
    return response.json();
  },
};

File: mobile_karyawan/lib/data/services/leave_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../../core/constants/app_constants.dart';
import '../../models/leave_model.dart';
import 'auth_service.dart';

class LeaveService {
  final String _baseUrl = AppConstants.baseUrl;
  final AuthService _authService = AuthService();

  Future<Map<String, String>> _getHeaders() async {
    final token = await _authService.getToken();
    return {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer $token',
    };
  }

  /// Ambil semua leave request milik user
  Future<List<LeaveRequest>> getLeaves() async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/leaves'),
        headers: await _getHeaders(),
      );
      if (response.statusCode == 200) {
        final List data = jsonDecode(response.body);
        return data.map((item) => LeaveRequest.fromJson(item)).toList();
      }
      return [];
    } catch (e) {
      print('Error fetching leaves: $e');
      return [];
    }
  }

  /// Ajukan cuti/izin baru
  Future<Map<String, dynamic>> applyLeave(Map<String, dynamic> data) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/leaves'),
      headers: await _getHeaders(),
      body: jsonEncode(data),
    );
    return jsonDecode(response.body);
  }

  /// Cek apakah hari ini user sedang cuti/izin yang sudah diapprove
  Future<LeaveRequest?> getActiveTodayLeave() async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/leaves/active-today'),
        headers: await _getHeaders(),
      );
      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        if (data == null) return null;
        return LeaveRequest.fromJson(data);
      }
      return null;
    } catch (e) {
      print('Error fetching active leave: $e');
      return null;
    }
  }
}
